### PR TITLE
Add mode fot topK/topKWeighed function to also include count/error statistics

### PR DESCRIPTION
**Pull request by UnamedRus**
### Changelog category (leave one):
- New Feature

Fix for https://github.com/ClickHouse/ClickHouse/issues/42029

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
topK/topKWeighed support mode, which return count of values and it's error.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

```
SELECT
    topK(10, 900, 'include_statistics')(number % 3000) AS top,
    top.1 AS key,
    tupleElement(top, 'count') AS cnt_arr,
    tupleElement(top, 'error') AS err_arr
FROM numbers_mt(10000000)
FORMAT Vertical

Query id: b6d68539-5cb1-4fa5-812c-bc331c7c61dc

Row 1:
──────
top:     [(581,3334,0),(160,3334,0),(666,3334,0),(175,3334,0),(305,3334,0),(378,3334,0),(124,3334,0),(794,3334,0),(833,3334,0),(688,3334,0)]
key:     [581,160,666,175,305,378,124,794,833,688]
cnt_arr: [3334,3334,3334,3334,3334,3334,3334,3334,3334,3334]
err_arr: [0,0,0,0,0,0,0,0,0,0]

SELECT
    topK(10, 900, 'include_statistics')((number % 300, CAST(number % 30, 'DateTime'), toString(number % 2))) AS top,
    top.1 AS key,
    tupleElement(top, 'count') AS cnt_arr,
    tupleElement(top, 'error') AS err_arr
FROM numbers_mt(10000000)
FORMAT Vertical

Query id: 1f4e1f10-fde7-4608-aed3-3bfa8187f8aa

Row 1:
──────
top:     [((5,'1970-01-01 00:00:05','1'),33334,0),((27,'1970-01-01 00:00:27','1'),33334,0),((85,'1970-01-01 00:00:25','1'),33334,0),((41,'1970-01-01 00:00:11','1'),33334,0),((20,'1970-01-01 00:00:20','0'),33334,0),((50,'1970-01-01 00:00:20','0'),33334,0),((80,'1970-01-01 00:00:20','0'),33334,0),((52,'1970-01-01 00:00:22','0'),33334,0),((2,'1970-01-01 00:00:02','0'),33334,0),((62,'1970-01-01 00:00:02','0'),33334,0)]
key:     [(5,'1970-01-01 00:00:05','1'),(27,'1970-01-01 00:00:27','1'),(85,'1970-01-01 00:00:25','1'),(41,'1970-01-01 00:00:11','1'),(20,'1970-01-01 00:00:20','0'),(50,'1970-01-01 00:00:20','0'),(80,'1970-01-01 00:00:20','0'),(52,'1970-01-01 00:00:22','0'),(2,'1970-01-01 00:00:02','0'),(62,'1970-01-01 00:00:02','0')]
cnt_arr: [33334,33334,33334,33334,33334,33334,33334,33334,33334,33334]
err_arr: [0,0,0,0,0,0,0,0,0,0]
```

Parameter naming & handling can be better, i guess.

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
This is an automated comment for commit a50ca1f42c82a25c163166412e55e716fec6d328 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/54508/a50ca1f42c82a25c163166412e55e716fec6d328/ci_running.html)
The overall status of the commit is 🔴 failure

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🔴 failure</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: docs/en/sql-reference/aggregate-functions/reference/topkweighted.md [hunk 1]
```diff
@@ -10,13 +10,20 @@ Returns an array of the approximately most frequent values in the specified colu
 **Syntax**
 
 ``` sql
-topKWeighted(N)(x, weight)
+topKWeighted(N)(column, weight)
+topKWeighted(N, load_factor)(column, weight)
+topKWeighted(N, load_factor, 'include_statistics')columnx, weight)
```
Comment (yakov-olkhovskiy) [inline docs/en/sql-reference/aggregate-functions/reference/topkweighted.md:15]:
```suggestion
topKWeighted(N, load_factor, 'include_statistics')(columnx, weight)
```

File: src/AggregateFunctions/AggregateFunctionTopK.cpp [hunk 2]
```diff
@@ -67,36 +68,36 @@ class AggregateFunctionTopKIPv4 : public AggregateFunctionTopK<DataTypeIPv4::Fie
...
-            params,
-            std::make_shared<DataTypeArray>(std::make_shared<DataTypeIPv4>()))
+            params)
     {}
 };
 
 
 template <bool is_weighted>
-IAggregateFunction * createWithExtraTypes(const DataTypes & argument_types, UInt64 threshold, UInt64 load_factor, const Array & params)
+IAggregateFunction * createWithExtraTypes(const DataTypes & argument_types, UInt64 threshold, UInt64 load_factor, UInt8 include_statistics, const Array & params)
```
Comment (yakov-olkhovskiy) [inline src/AggregateFunctions/AggregateFunctionTopK.cpp:83]: Why include_statistics is UInt8 and not a bool?

File: src/AggregateFunctions/AggregateFunctionTopK.cpp [hunk 3]
```diff
@@ -116,12 +117,13 @@ AggregateFunctionPtr createAggregateFunctionTopK(const std::string & name, const
 
     UInt64 threshold = 10;  /// default values
     UInt64 load_factor = 3;
+    UInt8  include_statistics = 0;
```
Comment (yakov-olkhovskiy) [inline src/AggregateFunctions/AggregateFunctionTopK.cpp:120]: the same - shouldn't it be bool?

**Review by yakov-olkhovskiy — COMMENTED** [review]
overall looks good but it would be nice to add some tests

File: src/AggregateFunctions/AggregateFunctionTopK.cpp (hunk 2 above)
Comment (UnamedRus) [reply src/AggregateFunctions/AggregateFunctionTopK.cpp:83]: Yes, will change to bool as well.

**Comment by UnamedRus** [comment]
> overall looks good but it would be nice to add some tests

It was more like POC example, will work on tests, if it's OK change.
I'm a little dislike, that change of parameter lead to change signature of function, but don't see any better way. (ie not use parameter for that)

Introduction of new aggregate function, like `TopKWeighedCount`, unfortunately will make people expect that they are different functions with different state and not same.


Anyway, it looks like, we better add 

APPROX_TOP_K (TopK) and APPROX_TOP_SUM (TopKWeighed) aliases, because they are much more common in other dbms.

https://github.com/ClickHouse/ClickHouse/issues/42029#issuecomment-1759769975

**Comment by yakov-olkhovskiy** [comment]
@UnamedRus could you please merge master again?

**Pull request merged**